### PR TITLE
Create sparse merkle tree

### PR DESCRIPTION
**Pull request by bun919tw**
Apply Sparse Merkle Tree (SMT) to block to generate merkle root.
Add `submit_block()` and `get_current_block()` function.

### CODE REVIEW

File: plasma_cash/utils/merkle/sparse_merkle_tree.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            prev_default = self.default_nodes[i - 1]
+            self.default_nodes.append(sha3(prev_default + prev_default))
+
+    def create_tree(self):
+        tree_level = self.leaves
+        for i in range(self.depth - 1):
+            next_level = {}
+            prev_index = -1
+            for index, value in tree_level.items():
+                if index % 2 == 0:
```
Comment (boolafish) [inline plasma_cash/utils/merkle/sparse_merkle_tree.py]: Although I am not a fan of using comment, would be nice here to tell what are these if else relationship in tree structure IMHO

File: plasma_cash/utils/merkle/sparse_merkle_tree.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class SparseMerkleTree(object):
+
+    def __init__(self, leaves={}):
+        self.depth = 256
+        self.leaves = OrderedDict(sorted(leaves.items(), key=lambda t: t[0]))
+        self.default_nodes = [b'\x00' * 32]
+        self.create_default_nodes()
+        if bool(leaves):
```
Comment (boolafish) [inline plasma_cash/utils/merkle/sparse_merkle_tree.py]: Minor: as discussed offline, could just use `if leaves`

File: plasma_cash/child_chain/child_chain.py [hunk 3]
```diff
@@ -10,14 +14,32 @@ def __init__(self, authority=plasma_config['AUTHORITY'],
...
-        new_owner = event['args']['depositor']
+        new_owner = utils.normalize_address(event['args']['depositor'])
         amount = event['args']['amount']
         uid = event['args']['uid']
         deposit_tx = Transaction(0, uid, amount, new_owner)
+        self.current_block.transaction_set[uid] = deposit_tx
+
+    def submit_block(self, sig):
+        signature = bytes.fromhex(sig)
+        assert signature != b'\x00' * 65
```
Comment (boolafish) [inline plasma_cash/child_chain/child_chain.py]: Minor: I think we can avoid assert and use proper exception in this package from the beginning

File: plasma_cash/child_chain/block.py [hunk 4]
```diff
@@ -21,6 +22,16 @@ def __init__(self, transaction_set=[], sig=b'\x00' * 65):
...
+    def merkilize_transaction_set(self):
+        hashed_transaction_dict = {tx.uid: tx.merkle_hash for tx in self.transaction_set}
+        self.merkle = SparseMerkleTree(hashed_transaction_dict)
+        return self.merkle.root
+
+    @property
+    def sender(self):
+        return get_sender(self.hash, self.sig)
+
     def sign(self, key):
```
Comment (boolafish) [inline plasma_cash/child_chain/block.py]: are we still using this?

File: plasma_cash/utils/merkle/sparse_merkle_tree.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            self.root = self.create_tree()
+        else:
+            self.root = self.default_nodes[self.depth - 1]
+
+    def create_default_nodes(self):
+        for i in range(1, self.depth):
```
Comment (boolafish) [inline plasma_cash/utils/merkle/sparse_merkle_tree.py]: Minor: could rename `i` to `level` for better readability

File: plasma_cash/utils/merkle/sparse_merkle_tree.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            prev_default = self.default_nodes[i - 1]
+            self.default_nodes.append(sha3(prev_default + prev_default))
+
+    def create_tree(self):
+        tree_level = self.leaves
+        for i in range(self.depth - 1):
```
Comment (boolafish) [inline plasma_cash/utils/merkle/sparse_merkle_tree.py]: Minor: same as above, could consider rename `i`

File: plasma_cash/utils/merkle/sparse_merkle_tree.py (hunk 2 above)
Comment (bun919tw) [reply plasma_cash/utils/merkle/sparse_merkle_tree.py]: Sure.

File: plasma_cash/utils/merkle/sparse_merkle_tree.py (hunk 1 above)
Comment (bun919tw) [reply plasma_cash/utils/merkle/sparse_merkle_tree.py]: You're right. I'll add some comments.

File: plasma_cash/child_chain/block.py (hunk 4 above)
Comment (bun919tw) [reply plasma_cash/child_chain/block.py]: Nope, I'll delete it.

File: plasma_cash/utils/merkle/sparse_merkle_tree.py (hunk 5 above)
Comment (bun919tw) [reply plasma_cash/utils/merkle/sparse_merkle_tree.py]: Great

**Pull request merged**